crates/bevy_audio/src/audio_source.rs

```diff
@@ -43,16 +46,84 @@ impl AssetLoader for Mp3Loader {
     }
 }
 
-pub trait Decodable: Send + Sync + 'static {
-    type Decoder;
+#[derive(Debug, Clone)]
+pub enum DecodableKind {
+    Decoder,
+    Repeat,
+}
 
-    fn decoder(&self) -> Self::Decoder;
+pub trait Decodable: Send + Sync + 'static {
+    fn decoder(&self) -> AudioSourceDecoder;
 }
 
 impl Decodable for AudioSource {
-    type Decoder = rodio::Decoder<Cursor<AudioSource>>;
+    fn decoder(&self) -> AudioSourceDecoder {
+        let decodable = rodio::Decoder::new(Cursor::new(self.clone())).unwrap();
+        match self.decodable_kind {
+            DecodableKind::Decoder => AudioSourceDecoder::Decoder(decodable),
+            DecodableKind::Repeat => AudioSourceDecoder::Repeat(decodable.repeat_infinite()),
+        }
+    }
+}
+
+impl AudioSource {
+    pub fn set_repeat_infinite(&mut self, repeat_infinite: bool) {
+        match repeat_infinite {
+            true => self.decodable_kind = DecodableKind::Repeat,
+            false => self.decodable_kind = DecodableKind::Decoder,
+        }
+    }
+}
+
+pub enum AudioSourceDecoder {
```

Comment:
I'm not convinced by this enum. It seems to result in lots of trivial match statements and a division that doesn't feel terribly natural to me.

Reply:
Yeah it is kind of just a stand alone boilerplate shell to `impl Iterator` and `impl Source`. It can definitely be rethought, especially the division between it and AudioSource. atm I don't have a any good ideas for getting around the trivial match statements / enum but I'll think abt it 🤔